src/Microsoft.ML.Data/DataLoadSave/PartitionedFileLoader.cs

```diff
@@ -0,0 +1,726 @@
+                }
+
+                return null;
+            }
+
+            public static bool TryParse(string str, out Column column)
+            {
+                column = null;
+
+                if (string.IsNullOrEmpty(str))
+                {
+                    return false;
+                }
+
+                if (!ColumnParsingUtils.TryParse(str, out string name, out string sourceStr, out string kindStr))
+                {
+                    return false;
+                }
+
+                DataKind? kind = null;
+                if (kindStr != null && TypeParsingUtils.TryParseDataKind(kindStr, out DataKind parsedKind, out KeyRange range))
+                {
+                    kind = parsedKind;
+                }
+
+                if (!int.TryParse(sourceStr, out int source))
+                {
+                    return false;
+                }
+
+                column = new Column()
+                {
+                    Name = name,
+                    Source = source,
+                    Type = kind
+                };
+
+                return true;
+            }
+
+            public bool TryUnparse(StringBuilder sb)
+            {
+                Contracts.AssertValue(sb);
+
+                sb.Append($"{Name}");
+
+                if (Type.HasValue)
+                {
+                    sb.Append($":{Type}");
+                }
+
+                sb.Append($":{Source}");
+
+                return true;
+            }
+        }
+
+        private readonly IHost _host;
+        private readonly IMultiStreamSource _files;
+        private readonly Column[] _columns;
+
+        // Number of tailing directories to include.
+        private readonly int _tailingDirCount;
+
+        // An underlying loader used on each individual loader.
+        private readonly SubComponent<IDataLoader, SignatureDataLoader> _subLoader;
+
+        private readonly IPartitionedPathParser _pathParser;
+
+        private const string RegistrationName = LoadName;
+
+        public PartitionedFileLoader(IHostEnvironment env, Arguments args, IMultiStreamSource files)
+        {
+            Contracts.CheckValue(env, nameof(env));
+            _host = env.Register(RegistrationName);
+            _host.CheckValue(args, nameof(args));
+            _host.CheckValue(files, nameof(files));
+
+            _pathParser = args.PathParserFactory.CreateComponent(_host);
+            _host.CheckValue(_pathParser, nameof(_pathParser), "Factory failed to create a FilePathSpec");
+
+            _subLoader = args.Loader;
+            _files = files;
+
+            string relativePath = GetRelativePath(args.BasePath, files);
+            _columns = ParseColumns(relativePath).ToArray();
+            _tailingDirCount = GetDirectoryCount(relativePath);
+
+            if (args.IncludePathColumn)
+            {
+                var pathCol = new Column()
+                {
+                    Name = "Path",
+                    Source = -1,
```

Comment:
> Source = -1, [](start = 19, length = 13)

any reason you want it to be int with -1 as "Not Specified" instead of nullable int?

Reply:
"-1" is not a "Not specified" value. "-1" is being used to signify a special column index relating to the FilePath. I've updated to a const variable to bring clarity.